packages/ember-application/lib/system/application.js

```diff
@@ -456,50 +577,74 @@ var Application = Namespace.extend(RegistryProxy, {
   },
 
   /**
-    Calling initialize manually is not supported.
-
-    Please see Ember.Application#advanceReadiness and
-    Ember.Application#deferReadiness.
+    Initialize the application and return a promise that resolves with the `Ember.Application`
+    object when the boot process is complete.
 
-    @private
-    @deprecated
-    @method initialize
-   **/
-  initialize() {
-    deprecate('Calling initialize manually is not supported. Please see Ember.Application#advanceReadiness and Ember.Application#deferReadiness');
-  },
+    Run any application initializers and run the application load hook. These hooks may
+    choose to defer readiness. For example, an authentication hook might want to defer
+    readiness until the auth token has been retrieved.
 
-  /**
-    Initialize the application. This happens automatically.
+    By default, this method is called automatically on "DOM ready"; however, if autoboot
+    is disabled, you would need to call this method yourself. Calling this method before
+    "DOM ready" is not recommended.
 
-    Run any initializers and run the application load hook. These hooks may
-    choose to defer readiness. For example, an authentication hook might want
-    to defer readiness until the auth token has been retrieved.
-
-    @private
-    @method domReady
+    @public
+    @method boot
+    @return {Promise<Ember.Application,Error>}
   */
-  domReady() {
-    if (this.isDestroyed) { return; }
+  boot() {
+    if (this._bootPromise) { return this._bootPromise; }
 
-    this.boot();
+    try {
+      this._bootSync();
+    } catch(_) {
+      // Ignore th error: in the asynchronous boot path, the error is already reflected
+      // in the promise rejection
+    }
 
-    return this;
+    return this._bootPromise;
   },
 
-  boot() {
-    if (this._bootPromise) { return this._bootPromise; }
+  /**
+    @private
+  */
+  _booted: false,
 
-    var defer = new Ember.RSVP.defer();
-    this._bootPromise = defer.promise;
-    this._bootResolver = defer;
+  /**
+    Unfortunately, a lot of existing code assumes the booting process is
+    "synchronous". Specifically, a lot of tests assumes the last call to
+    `app.advanceReadiness()` or `app.reset()` will result in the app being
+    fully-booted when the current runloop completes.
 
-    this.runInitializers();
-    runLoadHooks('application', this);
+    We would like new code (like the `visit` API) to stop making this assumption,
+    so we created the asynchronous version above that returns a promise. But until
+    we have migrated all the code, we would have to expose this method for use
+    *internall* in places where we need to boot an app "synchronously".
 
-    this.advanceReadiness();
+    @private
+  */
+  _bootSync() {
```

Comment:
The "detail" is important and is the primary reason why this method exits, though.

I wrote about it in the commit message: https://github.com/chancancode/ember.js/commit/1a423865c464847539a7f9535a49a2983d626fb4

Basically, I assumed we can just make boot async (the current code on master actually assumes that: the original `ember-application-visit` patch originally introduced `_bootPromise`), so I just went ahead and made all the things proper-async, then lost a ton of time figuring out why tests a failing and eventually realized that we actually **depend** on it being sync (in our tests, at least). See the comment above for the rest of the story.

So, there are definitely other ways to structure this, but my primary goal here is to make that implicit dependency clear and make sure others won't 1) waste time 2) have to understand the constraints before changing the semantics here (i.e .they won't change it by accident).

If you think that these are not that important, or the current implementation didn't quite achieve what I said above anyway (quite possible), then I'm happy to roll it back.